internal/cli/serve.go

```diff
@@ -24,23 +27,69 @@ func runServe(cmd *cobra.Command, cmdArgs []string) error {
 		"groups": runSettings.LockGroups,
 	}).Debug("lock groups")
 
-	logrus.WithFields(logrus.Fields{
-		"address": runSettings.ServiceAddress,
-		"port":    runSettings.ServicePort,
-	}).Info("starting service")
-
 	if runSettings == nil {
 		return errors.New("nil runSettings")
 	}
 	airlock := server.Airlock{*runSettings}
 
-	http.Handle(server.PreRebootEndpoint, airlock.PreReboot())
-	http.Handle(server.SteadyStateEndpoint, airlock.SteadyState())
+	stopCh := make(chan os.Signal)
+	signal.Notify(stopCh, os.Interrupt)
+	signal.Notify(stopCh, syscall.SIGTERM)
 
-	listenAddr := fmt.Sprintf("%s:%d", runSettings.ServiceAddress, runSettings.ServicePort)
-	if err := http.ListenAndServe(listenAddr, nil); err != nil {
-		return err
+	var statusService http.Server
+	if runSettings.StatusEnabled {
+		logrus.WithFields(logrus.Fields{
+			"address": runSettings.StatusAddress,
+			"port":    runSettings.StatusPort,
+		}).Info("status service")
+		go runStatusService(stopCh, statusService, airlock)
+		defer statusService.Close()
+	} else {
+		logrus.Warn("status service disabled")
 	}
 
+	var mainService http.Server
+	logrus.WithFields(logrus.Fields{
+		"address": runSettings.ServiceAddress,
+		"port":    runSettings.ServicePort,
+	}).Info("main service")
+	go runMainService(stopCh, mainService, airlock)
+	defer mainService.Close()
+
+	<-stopCh
 	return nil
 }
+
+// runMainService runs the main HTTP service
+func runMainService(stopCh chan os.Signal, service http.Server, airlock server.Airlock) {
```

Comment:
Any reason to not combine `runMainService` & `runStatusService`? They're largely the same thing (the only differences really being the `mux.Handle`'s in `runMainService` & the service name in the error. Should be pretty simple to accept a hashmap that contains the endpoints and their handler funcs and iterate through them.

Reply:
Do you mean on the same port? I split them on purpose as the status service is usually locally consumed (by the monitoring infrastructure, i.e. Prometheus and Kubernetes probes) while the main service is exposed to external consumers.

Reply:
No, I meant in the sense of having a generic `runService` function that is called in the `runServe` calls (e.x.: instead of `go runStatusService(stopCh, statusService, airlock)` you'd have something like:
```
handlerMap := make(map[string]func() http.Handler)
handlerMap[status.MetricsEndpoint] = status.Metrics
go runService(stopCh, statusService, airlock, handlerMap)
```

where `runService` looks like:

```
func runService(stopCh chan os.Signal, service http.Server, airlock server.Airlock, handlerMap map[string]func() http.Handler) {
    ...
    mux := http.NewServeMux()
    for endpoint, f := range handlerMap {
        mux.Handle(endpoint, f)
    }
    ...
}
```

Reply:
Ack. Sorry I misunderstood your initial suggestion.